---
title: "iSeq *Escherichia coli* and *Listeria monocytogenes* Multi-laboratory Validation"
subtitle: "iSeq Lab: STATE  \nRun Number: 1  \nRun Date:   \nReport Date: `r format(Sys.time(), '%B %d, %Y')`"
output:
  word_document: 
    reference_docx: "/workdir/iSeq_Ecoli/report_templates/testreport_template.docx"
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(flextable)
library(officer)
library(magrittr)
ReadMetrics<-read.table("ReadMetrics_DATETIME.tsv",header = T,sep = "\t",check.names=T)
ReadMetrics$Isolate<-substr(ReadMetrics$Isolate,14,18)
AssemblyMetrics<-read.table("AssemblyMetrics_DATETIME.tsv",header = T,sep = "\t",check.names=T)
AssemblyMetrics$Isolate<-substr(AssemblyMetrics$Isolate,14,18)
AlignmentMetrics<-read.table("AlignmentMetrics_DATETIME.tsv",header = T,sep = "\t",check.names=T)
AlignmentMetrics$Isolate<-substr(AlignmentMetrics$Isolate,14,18)
AnnotationTable<-read.table("AnnotationTable_DATETIME.tsv",header = T,sep = "\t",check.names=T)
AnnotationTable$Isolate<-substr(AnnotationTable$Isolate,14,18)
SubtypingTable<-read.table("SubtypeTable_DATETIME.tsv",header=T, sep="\t",check.names=T)
SubtypingTable$Isolate<-substr(SubtypingTable$Isolate,14,18)
SoftwareTable<-read.table("Software_table_DATETIME.tsv", header=F, sep="\t")

```
## Introduction
The Veterinary Laboratory Investigation and Response Network (Vet-LIRN) provided samples to thirteen pilot laboratories for assessment of the utility of the Illumina iSeq platform for bacterial surveillance. The results from the sequencing of these samples will allow Vet-LIRN and the iSeq laboratories the opportunity to assess their capability to test previously characterized isolates and confirm that harmonized methods are working properly within the laboratories. Throughout this report, values that are either unexpected or outside of current GenomeTrakr benchmarks are colored in yellow or red.

## Background
Samples consisted of four *Escherichia coli* and two *Listeria monocytogenes* isolates collected and previously distributed by Genome Trackr for MiSeq proficiency testing. Laboratories extracted DNA from these isolates and prepared sequencing libraries using Nextera DNA Flex, which were then run on the iSeq platform. Read quality, *de novo* assembly metrics, reference mapping, subtyping, and gene annotation were assessed for each sample.

## Conclusion
TODO

##### 

## Read Metrics
The table below gives quality metrics for the reads produced by this run. Targets are average forward and reverse quality scores >=30 for both the forward and reverse reads and coverage depth >=40X  for *E. coli* and >=20X for *L. monocytogenes*. Coverage depth is estimated by multiplying the number of reads by the average read length and dividing by the expected genome length (5 Mbp for *E. coli* and 3 Mbp for *L. monocytogenes*). 
```{r read table, echo=FALSE}

flextable(ReadMetrics) %>% 
  color( ~ F.Q < 30, color = "darkred", ~ F.Q) %>%
  color( ~ R.Q < 30, color = "darkred", ~ R.Q) %>%
  color( ~ as.numeric(substr(Est..Coverage,0,nchar(as.character(Est..Coverage))-1))<40 & substr(Isolate,1,2)=="EC", color = "darkred", ~ Est..Coverage) %>%
  color( ~ as.numeric(substr(Est..Coverage,0,nchar(as.character(Est..Coverage))-1))<20 & substr(Isolate,1,2)=="LM", color = "darkred", ~ Est..Coverage) %>%
  set_header_labels(F.Length = "R1 Length", R.Length= "R2 Length",F.Q="R1 Quality", R.Q = "R2 Quality", Est..Coverage = "Est. Depth") %>%
  bold(part="header") %>%
  autofit() %>% align(align="center", part="all")
```
  
\   \   
  

## Assembly Metrics
The table below describes the *de novo* assemblies generated for each isolate by SKESA. Target values for the number of contigs and total assembly length are are <=600 and 4.2--5.9 Mbp for *E. coli* and <=200 and 2.8--3.1 Mbp for *L. monocytogenes*. Values outside of these ranges are colored in red. N50/NG50 values below 10,000 indicate a highly fragmented assembly and are colored in yellow. Unaligned bases, also in yellow, are suggestive of a sample swap or contamination. The aligned percentage shows the proportion of the reference genome covered by the *de novo* assembly. 
```{r assembly table, echo=FALSE}

flextable(AssemblyMetrics) %>%
  color( ~ N50 < 10000 | NG50 < 10000, color = "goldenrod", ~ N50 + NG50) %>%
  color( ~ Unaligned.Length > 0, color = "goldenrod", ~ Unaligned.Length) %>%
  color( ~ Contigs > 200 &  substr(Isolate,1,2)=="LM", color = "darkred", ~ Contigs) %>%
  color( ~ Contigs > 600 &  substr(Isolate,1,2)=="EC", color = "darkred", ~ Contigs) %>%
  color( ~ Total.Length > 3100000 &  substr(Isolate,1,2)=="LM", color = "darkred", ~ Total.Length) %>%
  color( ~ Total.Length > 5900000 &  substr(Isolate,1,2)=="EC", color = "darkred", ~ Total.Length) %>%
  color( ~ Total.Length < 2800000 &  substr(Isolate,1,2)=="LM", color = "darkred", ~ Total.Length) %>%
  color( ~ Total.Length < 4200000 &  substr(Isolate,1,2)=="EC", color = "darkred", ~ Total.Length) %>%
  set_header_labels(Total.Length = "Length", Genome.Fraction....="Aligned Pct.", Unaligned.Length="Unaligned Bases") %>%
  bold(part="header") %>%
  autofit() %>% align(align="center", part="all")
```

##### 
## Alignment Metrics
The table below shows result from mapping reads from each isolate against the corresponding closed reference genome using the CFSAN SNP Pipeline. Values outside of the expected ranges are colored in yellow. Low mapping percentages and high SNP counts suggest a sample swap or contamination. 
```{r aligment table, echo=FALSE}

flextable(AlignmentMetrics) %>%
  color( ~ SNPs >10, color = "goldenrod", ~ SNPs) %>%
  color( ~ Reads.Mapped.... < 97.5, color = "goldenrod", ~ Reads.Mapped....) %>%
  color( ~ Insert.Size > 350| Insert.Size < 300, color = "goldenrod", ~ Insert.Size) %>%
  set_header_labels(Reads.Mapped.... = "Reads Mapped (%)", Insert.Size="Insert Size", Read.Depth="Read Depth") %>%
  bold(part="header") %>%
  autofit() %>% align(align="center", part="all")
```
  
\  \   
  

## Subtyping
This table shows the serotype of *E. coli* isolates and the 7-gene multilocus sequence type (ST) of *E. coli* and *L. monocytogenes* isolates. Serotypes and STs that differ from the correspdoning reference genome are colored in red.
```{r subtyping table, echo=FALSE}

flextable(SubtypingTable) %>%
  bold(part="header") %>%
  set_header_labels(O.type = "O-type", H.type="H-type") %>%
  color( ~ O.type != c(rep("O157",4),rep("NA",2)), color="darkred", ~ O.type) %>%
  color( ~ H.type != c(rep("H7",4),rep("NA",2)), color="darkred", ~ H.type) %>%
  color( ~ ST != c(rep(11,4),3,7), color="darkred", ~ ST + Alleles) %>%
  autofit() %>% align(align="center", part="all")
```

##### 
## Gene Annotations
This table compares the AMR genes (ARGs) and virulence factors (VFs) detected in each isolate compared to those detected in the corresponding reference genome. Virulence factor screening was not conducted for *L. monocytogenes*.
```{r annotation table, echo=FALSE}
flextable(AnnotationTable) %>%
  bold(part="header") %>%
  set_header_labels(Missed.ARGs = "Missed ARGs", Unexpected.ARGs="Unexpected ARGs",Missed.VFs = "Missed VFs", Unexpected.VFs="Unexpected VFs") %>%
  color(~ Missed.ARGs != "None", color = "darkred", ~ Missed.ARGs) %>%
  color(~ !Missed.VFs %in% c("None","NA"), color = "darkred", ~ Missed.VFs) %>%
  color(~ Unexpected.ARGs != "None", color = "darkred", ~ Unexpected.ARGs) %>%
  color(~ !Unexpected.VFs %in% c("None","NA"), color = "darkred", ~ Unexpected.VFs) %>%
  autofit() %>% align(align="center", part="all")
```
  
 \  \  
  

## Software Versions
   
   
```{r software table, echo=FALSE}

flextable(SoftwareTable) %>%
 delete_part(part="header") %>% 
 bg(bg = "grey85", j=1) %>%
 border_outer(part="body", border=fp_border(width=1)) %>%
 border_inner_h(part="body", border=fp_border(width=1)) %>%
 border_inner_v(part="body", border=fp_border(width=1)) %>%
 autofit() %>% align(align="left", part="all")
```
  
\  \   
  

**Report prepared by:**  
**Report approved by:**  
